lib/optimizely/odp/odp_manager.rb

```diff
@@ -0,0 +1,144 @@
+        return
+      end
+
+      unless @segment_manager
+        segments_cache ||= LRUCache.new(
+          Helpers::Constants::ODP_SEGMENTS_CACHE_CONFIG[:DEFAULT_CAPACITY],
+          Helpers::Constants::ODP_SEGMENTS_CACHE_CONFIG[:DEFAULT_TIMEOUT_SECONDS]
+        )
+        @segment_manager = Optimizely::OdpSegmentManager.new(segments_cache, nil, @logger)
+      end
+
+      @event_manager ||= Optimizely::OdpEventManager.new(logger: @logger)
+
+      @segment_manager.odp_config = @odp_config
+      @event_manager.start!(@odp_config)
+    end
+
+    def fetch_qualified_segments(user_id:, options:)
+      # Returns qualified segments for the user from the cache or the ODP server if not in the cache.
+      #
+      # @param user_id - The user id.
+      # @param options - An array of OptimizelySegmentOptions used to ignore and/or reset the cache.
+      #
+      # @return - Array of qualified segments or nil.
+      options ||= []
+      unless @enabled
+        @logger.log(Logger::ERROR, ODP_LOGS[:ODP_NOT_ENABLED])
+        return nil
+      end
+
+      if @odp_config.odp_state == ODP_CONFIG_STATE[:UNDETERMINED]
+        @logger.log(Logger::ERROR, 'Cannot fetch segments before the datafile has loaded.')
+        return nil
+      end
+
+      @segment_manager.fetch_qualified_segments(ODP_MANAGER_CONFIG[:KEY_FOR_USER_ID], user_id, options)
+    end
+
+    def identify_user(user_id:)
+      unless @enabled
+        @logger.log(Logger::DEBUG, 'ODP identify event is not dispatched (ODP disabled).')
+        return
+      end
+
+      case @odp_config.odp_state
+      when ODP_CONFIG_STATE[:UNDETERMINED]
+        @logger.log(Logger::DEBUG, 'ODP identify event is not dispatched (datafile not ready).')
+        return
+      when ODP_CONFIG_STATE[:NOT_INTEGRATED]
+        @logger.log(Logger::DEBUG, 'ODP identify event is not dispatched (ODP not integrated).')
+        return
+      end
+
+      @event_manager.send_event(
+        type: ODP_MANAGER_CONFIG[:EVENT_TYPE],
+        action: 'identified',
+        identifiers: {ODP_MANAGER_CONFIG[:KEY_FOR_USER_ID] => user_id},
+        data: {}
+      )
+    end
+
+    def send_event(type:, action:, identifiers:, data:)
+      # Send an event to the ODP server.
+      #
+      # @param type - the event type.
+      # @param action - the event action name.
+      # @param identifiers - a hash for identifiers.
+      # @param data - a hash for associated data. The default event data will be added to this data before sending to the ODP server.
+      unless @enabled
+        @logger.log(Logger::ERROR, ODP_LOGS[:ODP_NOT_ENABLED])
+        return
+      end
+
+      unless Helpers::Validator.odp_data_types_valid?(data)
+        @logger.log(Logger::ERROR, ODP_LOGS[:ODP_INVALID_DATA])
+        return
+      end
+
+      @event_manager.send_event(type: type, action: action, identifiers: identifiers, data: data)
+    end
+
+    def update_odp_config(api_key, api_host, segments_to_check)
```

Comment:
This is great! I am just curios who will call this function? should we add a notification listener here and make odpmanager self update itself? Or are we expecting to do something like that and trigger updates from the outside by calling this method?

Reply:
Yea the idea is to register the listener at client initialization and have the callback be a client method, thus avoiding having to pass in the `NotificationListener` and the `ProjectConfig`. Seems to make sense because the callback needs access to both `ProjectConfig` and `OdpConfig`. Thoughts/opinions?